openedx/openedx-platform
# fix: upgrade py2neo from 3.1.2 to 2021.1.5

## Description
The dump_to_neo4j management command has not been working
since the upgrade to python 3.8. The latest version of
python that py2neo 3.1.2 states support for is python 3.5,
so this isn't surprising.

The earliest non-prerelease version of py2neo that supports
python 3.8 is 2020.x (skipping the 4.x and 5.x series). Since
we're going as far as a 2020.x, we may as well upgrade all the
way to the newest series, 2021.x. This commit does that upgrade,
as well as a handful of minor code modifications in order
to handle breaking changes that have been made to the py2neo
API.

This will also require an upgrade of Coursegraph's Neo4j
version from 3.2.x to 3.5.x.

## Supporting information

Ticket: https://openedx.atlassian.net/browse/TNL-8386
Configuration PR to update the deployed neo4j version: https://github.com/edx/configuration/pull/6502
Devstack PR to update the dev neo4j version: https://github.com/edx/devstack/pull/824

## Testing instructions

Check out this branch of edx-platform. Check out [this](https://github.com/edx/devstack/pull/824) branch of devstack.

In devstack, `make lms-shell`, and `make requirements` in order to install the new py2neo version.

In devstack, destroy and reprovision coursegraph, ensuring it completes successfully:
```
make dev.destroy.coursegraph  # Can be skipped if you've never provisioned it before.
make dev.provision.coursegraph 
```

Visit coursegraph at http://localhost:7474. Run a sanity-checking query or two, such as:
```
MATCH (c:course)-[:PARENT_OF*]->(v:vertical)
WHERE c.org = 'edX'
RETURN c.display_name, c.course_key, v.display_name, v.location
```

## Deadline

Sooner is better than later.
